File: 0017-Letter-Combinations-of-a-Phone-Number/lc0017.cc
// C++ solution.

class Solution {
public:
  vector<string> letterCombinations(string digits) {
    int sz = digits.size();
    if (sz > 0) {
      res = string(sz, '0');
      backtracking(0, sz, digits);
    }
    return ans;
  }

  void backtracking(int depth, const int sz, string& digits) {
    if (depth == sz) {
      ans.push_back(res);
      return;
    }

    int button = digits[depth] - '0';
    for (int i = 0; i < dig[button].size(); i++) {
      res[depth] = dig[button][i];
      backtracking(depth+1, sz, digits);
    }
  }

private:
  string dig[10] = {"", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
  vector<string> ans;
  string res;
};

File: 0017-Letter-Combinations-of-a-Phone-Number/lc0017.py
# Python solution.

class Solution:
    def __init__(self):
        self.dig = ["", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]
        self.ans = []
        self.res = []

    def letterCombinations(self, digits: str) -> List[str]:
        sz = len(digits)
        self.digits = digits
        if sz > 0:
            self.res = ['0' for i in range(sz)]
            self.backtracking(0, sz)
        return self.ans

    def backtracking(self, depth: int, sz: int) -> None:
        if depth == sz:
            self.ans.append("".join(self.res))
            return

        button = int(self.digits[depth])
        for letter in self.dig[button]:
            self.res[depth] = letter
            self.backtracking(depth+1, sz)
